### PR TITLE
PR: Add search paths to PATH in is_program_installed

### LINKED ISSUE #14136: Conda Environment Status Shows Full Path
## Problem Description
@steff456 
Environment status shows full path to conda environment, rather than just environment name.
Pyenv environment status shows correctly.

Screen Shot 2020-11-03 at 8 40 33 AM
Screen Shot 2020-11-03 at 8 40 51 AM

## Versions

* Spyder version: 4.2.0.dev0 0dfbe8967
* Python version: 3.7.9 64-bit
* Qt version: 5.9.6
* PyQt5 version: 5.9.2
* Operating System: Darwin 19.6.0

### Dependencies

```

# Mandatory:
applaunchservices >=0.1.7      :  0.2.1 (OK)
atomicwrites >=1.2.0           :  1.4.0 (OK)
chardet >=2.0.0                :  3.0.4 (OK)
cloudpickle >=0.5.0            :  1.6.0 (OK)
diff_match_patch >=20181111    :  20200713 (OK)
intervaltree >=3.0.2           :  3.1.0 (OK)
IPython >=4.0                  :  7.18.1 (OK)
jedi =0.17.2                   :  0.17.2 (OK)
jsonschema >=3.2.0             :  3.2.0 (OK)
nbconvert >=4.0                :  6.0.7 (OK)
numpydoc >=0.6.0               :  1.1.0 (OK)
parso =0.7.0                   :  0.7.0 (OK)
pexpect >=4.4.0                :  4.8.0 (OK)
pickleshare >=0.4              :  0.7.5 (OK)
psutil >=5.3                   :  5.7.2 (OK)
pygments >=2.0                 :  2.7.2 (OK)
pylint >=1.0                   :  2.6.0 (OK)
pyls >=0.35.0;<1.0.0           :  test-selector-01+92.g0dfbe8967 (OK)
pyls_black >=0.4.6             :  0.4.6 (OK)
pyls_spyder >=0.1.1            :  0.1.1 (OK)
qdarkstyle >=2.8               :  2.8.1 (OK)
qtawesome >=0.5.7              :  1.0.1 (OK)
qtconsole >=4.7.7              :  4.7.7 (OK)
qtpy >=1.5.0                   :  1.9.0 (OK)
rtree >=0.8.3                  :  0.9.4 (OK)
setuptools >=39.0.0            :  50.3.0.post20201006 (OK)
sphinx >=0.6.6                 :  3.2.1 (OK)
spyder_kernels >=1.9.4;<1.10.0 :  1.10.0.dev0 (OK)
three_merge >=0.1.1            :  0.1.1 (OK)
watchdog >=0.10.3              :  0.10.3 (OK)
zmq >=17                       :  19.0.2 (OK)

# Optional:
cython >=0.21                  :  0.29.21 (OK)
matplotlib >=2.0.0             :  3.3.2 (OK)
numpy >=1.7                    :  1.19.2 (OK)
pandas >=1.1.1                 :  1.1.3 (OK)
scipy >=0.17.0                 :  1.5.2 (OK)
sympy >=0.7.3                  :  1.6.2 (OK)
```

### PR DESCRIPTION
**Pull request by mrclary**
## Description of Changes
Temporarily add required paths to `sys.environ['PATH']` in `spyder.utils.programs.is_program_installed`.

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

By adding platform dependent paths to `os.environ['PATH']`, developers will not have to consider macOS application cases when using `find_program`. Additionally, developers will not have to consider cases in which Spyder is not launched from login shells (i.e. user profiles are not sourced).

Since `PATH` is only temporarily modified to locate a given program, it should not affect the behavior of Spyder or Consoles.

Developers should update the `req_paths` in the respective platform control sequence to include paths in the search for programs.

### Issues Resolved
Fixes #14136 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
mrclary

### CODE REVIEW

**Comment by mrclary** [comment]
@ccordoba12 @steff456, per conversation [here](https://github.com/spyder-ide/spyder/pull/13950#discussion_r516201754), after merging #13950, this PR can be rebased and reviewed.

**Comment by jitseniesen** [comment]
Why is it necessary to change the environment variable PATH at all? Would it be possible instead to change
```
for path in os.environ["PATH"].split(os.pathsep):
```
to
```
for path in os.environ["PATH"].split(os.pathsep) + req_paths:
```
or something like that?

**Comment by mrclary** [comment]
> Why is it necessary to change the environment variable PATH at all? Would it be possible instead to change
> 
> ```
> for path in os.environ["PATH"].split(os.pathsep):
> ```
> 
> to
> 
> ```
> for path in os.environ["PATH"].split(os.pathsep) + req_paths:
> ```
> 
> or something like that?

@jitseniesen, yours is a much better idea. Thanks.

File: spyder/utils/programs.py [hunk 1]
```diff
@@ -71,17 +71,50 @@ def get_temp_dir(suffix=None):
...
+            osp.join('/usr', 'local', 'bin'),
+            osp.join(home, 'opt', 'anaconda3', 'condabin'),
+            osp.join(home, 'opt', 'miniconda3', 'condabin'),
+            osp.join('/opt', 'anaconda3', 'condabin'),
+            osp.join('/opt', 'miniconda3', 'condabin')
+        ])
+    elif sys.platform.startswith('linux'):
+        # TODO: what are the possible conda paths?
+        # TODO: what are the possible pyenv paths?
+        pass
```
Comment (ccordoba12) [inline spyder/utils/programs.py:111]:
```suggestion
        req_paths.extend([
            osp.join('/usr', 'local', 'bin'),
            osp.join(home, 'anaconda3', 'condabin'),
            osp.join(home, 'miniconda3', 'condabin'),
            osp.join('/opt', 'anaconda3', 'condabin'),
            osp.join('/opt', 'miniconda3', 'condabin')
        ])
```

File: spyder/utils/programs.py [hunk 2]
```diff
@@ -71,17 +71,50 @@ def get_temp_dir(suffix=None):
...
+    Darwin:
+        /usr/local/bin
+        ~/opt/anaconda3/condabin
+        ~/opt/miniconda3/condabin
+        /opt/anaconda3/condabin
+        /opt/miniconda3/condabin
+    Linux:
+        ?
+    Windows:
+        ?
```
Comment (ccordoba12) [inline spyder/utils/programs.py:92]:
```suggestion
```

Default paths are present in the code below, so I see no need to add them here.

File: spyder/utils/programs.py [hunk 3]
```diff
@@ -71,17 +71,50 @@ def get_temp_dir(suffix=None):
...
+            osp.join('/opt', 'miniconda3', 'condabin')
+        ])
+    elif sys.platform.startswith('linux'):
+        # TODO: what are the possible conda paths?
+        # TODO: what are the possible pyenv paths?
+        pass
+    elif WINDOWS:
+        # TODO: what are the possible conda paths?
+        # TODO: what are the possible pyenv paths?
+        pass
```
Comment (ccordoba12) [inline spyder/utils/programs.py:115]: @dalthviz, could you provide the default installation locations for Anaconda and Miniconda on Windows here?

**Review by ccordoba12 — COMMENTED** [review]
Thanks @mrclary, this looks pretty good. I left an initial review for you.

File: spyder/utils/programs.py (hunk 3 above)
Comment (dalthviz) [reply spyder/utils/programs.py:115]: For Anaconda:

* Installations for just the current user: the path is something like `C:\Users\<user>\anaconda3\condabin`
* Installations for all the user: the default path is something like `C:\ProgramData\Anaconda3\condabin`

For Miniconda:

* Installations for just the current user: the path is something like `C:\Users\<user>\miniconda3\condabin`
* Installations for all the user: the default path is something like `C:\ProgramData\Miniconda3\condabin`

Comment (mrclary) [reply spyder/utils/programs.py:115]: Looks like standard installation location for `pyenv` is `$HOME\.pyenv\pyenv-win\bin` and for `conda` is 
`$HOME\Anaconda3\condabin`
`$HOME\Miniconda3\condabin`
`C:\Anaconda3\condabin`
`C:\Miniconda3\condabin`
`C:\ProgramData\Anaconda3\condabin`
`C:\ProgramData\Miniconda3\condabin`

@dalthviz, can you confirm? Additions? Subtractions?

Comment (dalthviz) [reply spyder/utils/programs.py:115]: LGTM :+1:

**Review by ccordoba12 — APPROVED** [review]
Looks good to me now, thanks @mrclary!

**Pull request merged**